library/src/main/java/com/davidstemmer/screenplay/flow/Screenplay.java

```diff
@@ -14,73 +14,94 @@
 import flow.Backstack;
 import flow.Flow;
 
+
+/**
+ * The Screenplay object handles the navigation logic for a Screenplay application.
+ */
 public class Screenplay implements Flow.Listener {
 
     private SceneState screenState = SceneState.NORMAL;
 
     private final Stage stage;
 
-    private final ArrayDeque<Scene> previousScenes = new ArrayDeque<Scene>();
+    private final ArrayDeque<Scene> previousBackstack = new ArrayDeque<>();
 
     public Screenplay(Stage stage) {
         this.stage = stage;
     }
 
     @Override
-    public void go(Backstack nextBackstack, Flow.Direction direction, Flow.Callback callback) {
+    public void go(Backstack backstack, Flow.Direction direction, Flow.Callback callback) {
 
-        Scene incomingScene = (Scene) nextBackstack.current().getScreen();
+        ArrayDeque<Scene> nextBackstack = deque(backstack);
 
-        ArrayDeque<Scene> incomingScenes;
-        ArrayDeque<Scene> outgoingScenes;
+        ArrayDeque<Scene> animatedScenesIn;
+        ArrayDeque<Scene> animatedScenesOut;
 
         SceneCut.Builder sceneCut = new SceneCut.Builder()
                 .setDirection(direction)
                 .setCallback(callback);
 
         if (direction == Flow.Direction.BACKWARD) {
-            Scene outgoingScene = previousScenes.getLast();
-            if (outgoingScene.isStacking()) {
-                incomingScenes = new ArrayDeque<>();
-            } else {
-                incomingScenes = getLastSceneBlock(nextBackstack);
-            }
-            outgoingScenes = new ArrayDeque<>();
-            outgoingScenes.push(outgoingScene);
+            ArrayDeque<Scene> difference = difference(previousBackstack, nextBackstack);
+            animatedScenesOut = getLastSceneBlock(difference);
+            animatedScenesIn = moveToNewSceneBlock(difference) ?
```

Comment:
it's not clear what this is asking, I'll figure it out when I get there, but I can't tell from this reading what it's basing the branch on.